### PR TITLE
Add preprocessing option letterbox for the classify task

### PR DESCRIPTION
**Pull request by timstokman**
This PR solves #3029. The default preprocessing for the classification task is very suboptimal for datasets with high aspect ratios or differing sizes, so I added the option `letterbox`.  If letterbox is set to `True` for the classification task, letterbox preprocessing will be used for the image dataset, which is more appropriate for datasets with high aspect ratios.

The default preprocessing step will stay exactly the same, to ensure it's not changing the precision of existing checkpoints.

It can be used through the CLI:

```
yolo classify train letterbox=True
yolo classify val letterbox=True
```

or through the python API:

```
model = ultralytics.YOLO(f'yolov8{args.model_size}-cls.pt')
model.train(letterbox=True)
```

See the discussion in #3029 as to why the current preprocessing doesn't work well for some datasets.

I didn't add this option for the other tasks because for the segment/detection/pose task the only sensible option is the current one. No need to make that configurable.

### Summary
📦📝🧪

This pull request adds a new `letterbox` argument to the `classify_transforms` function and the `ClassificationDataset` and `BasePredictor` classes, to enable letterbox preprocessing for images with uneven aspect ratios in classify mode. It also updates the documentation and the tests to reflect this change and clarify the role of the existing `rect` argument.

> _Oh we are the ultralytics crew, we code with skill and pride_
> _We `letterbox` our images, no matter how they're sized_
> _We `rect` them and we crop them, we make them fit the task_
> _We `classify_transforms` with ease, and that's all that we ask_

### Walkthrough
*  Add `letterbox` argument and modify `rect` argument for classify mode ([link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-4a4a3691b08dc4cb045c88d4405e25fe3e154c35ddf71370e94e336b5f3b417dL198-R198), [link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-4a4a3691b08dc4cb045c88d4405e25fe3e154c35ddf71370e94e336b5f3b417dR224), [link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-732d61cb4969eff986cec58058000b2949ea3b429a97ca5a1045897e754145c5L28), [link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-732d61cb4969eff986cec58058000b2949ea3b429a97ca5a1045897e754145c5R40-R41), [link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-8945a175b39a1efcabd4d2fb09b583d2f25409f05038ca9a0f1165e39ee48528L950-R958), [link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-0834dc62a5ebc93f43947ef49530b3e973013f844888646e5fb9962bfefecd90L228-R228), [link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-9fa2bf7f1273283680da3daeb8bae8e20cf8d477f838575796d2bf3b415bc534L211-R214))
* Add test function for `classify_transforms` function ([link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-882061b431ed3670f5b4a045ae0f4e1f140cc785db77fe77585b820bdec6f73dL3-R9), [link](https://github.com/ultralytics/ultralytics/pull/3116/files?diff=unified&w=0#diff-882061b431ed3670f5b4a045ae0f4e1f140cc785db77fe77585b820bdec6f73dR134-R153))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved image preprocessing for classification in Ultralytics YOLO.

### 📊 Key Changes
- Updated `rect` flag documentation to specify it is for "classify train only."
- Added a new `letterbox` flag for letterbox preprocessing in classification training.
- Extended the classification transform function to include the `letterbox` option.
- Integrated `letterbox` and `rect` flags into dataset preprocessing.
- Adapted the predictor setup to apply new `letterbox` preprocessing options.

### 🎯 Purpose & Impact
- 🎨 Ensures more flexible image preprocessing for various aspect ratios, potentially enhancing model performance on diverse datasets.
- 🔍 Promotes better understanding of the `rect` flag through clearer documentation.
- 📐 Provides options for maintaining aspect ratios via letterbox padding during preprocessing, which may be beneficial when training on images with inconsistent sizes or shapes.
- 🧠 These changes could lead to more accurate classification models by improving how they handle input images with varying aspect ratios.

### CODE REVIEW

**Comment by glenn-jocher** [comment]
@timstokman thanks for the PR! Do you have ImageNet results before/after? What about results on the different preprocessing options in the new argument?

A challenge is that we need arguments to apply broadly ideally, i.e. imgsz is used for train/val/predict, so if we have an argument called preprocessing it should in theory also apply to det/seg/pose. We might need to rename to preprocessing_cls. What do you think?

**Comment by glenn-jocher** [comment]
@timstokman @sergiuwaxmann classify preprocessing tests using this PR (with all options) vs main in https://wandb.ai/glenn-jocher/exp-preprocessing

**Comment by timstokman** [comment]
I don't have the imagenet results, only tested this on my own dataset. Given that imagenet has reasonable (mostly square) aspect-ratios, I don't think the preprocessing option will actually matter much one way or the other for that specific dataset. Still.. looking forward to the results of the test you just started, might show some interesting results

Renaming the option to preprocessing_cls makes sense. If needed I can expand the PR to have the preprocessing option apply to other tasks, but for non-classification tasks I don't think a preprocessing option actually makes sense for detection/segmentation/pose tasks (centercrop will remove information, resize warps aspect ratios).

I'll do the rename for now.

**Comment by timstokman** [comment]
@glenn-jocher There are other options that are task-specific which can be used in the command-line tool. None of these have naming to suggest they're specific to one task (like `dropout` which only applies to the classify task). The naming you suggested makes sense, it might also make sense to create a separate PR to rename the other task-specific options at some point.

**Comment by timstokman** [comment]
The name of the option is still the blocker for merging this I'm guessing?

The core problem is that:

* The name of the option didn't include an indication that it was task specific, so @glenn-jocher made suggestion to rename it here: https://github.com/ultralytics/ultralytics/pull/3116#issuecomment-1586687674
* Except there are more options (like `mask_ratio` and `dropout`) which are also class-specific. So that rename makes the option name `preprocessing_cls` a bit inconsistent.

I see three ways forward:

* I'll stick with the option name `preprocessing` instead of `preprocessing_cls`, that way it's consistent with the naming of other options. 
* The `preprocessing_cls` name is fine, the other options can be renamed when needed
* I can add a postfix name to all task-specific options. This might break things for a number of integrations, so I'd prefer not to do that.

@glenn-jocher Which of these three do you prefer? I'd rather this PR gets merged at some point, so please let me know what is blocking it.


PS. The experiment was here was pretty interesting: https://wandb.ai/glenn-jocher/exp-preprocessing , though not very unexpected. These options were added to improve the performance on datasets with large aspect-ratios. For imagenet that's not the case.

**Comment by glenn-jocher** [comment]
@timstokman hi there,

Thank you for your feedback and for bringing up the issue regarding the option name. The main blocker for merging this PR is indeed the name of the option. The original suggestion was to rename it to `preprocessing_cls` to indicate that it is task-specific. However, you correctly pointed out that there are other options like `mask_ratio` and `dropout` which are also task-specific but don't have specific naming.

Moving forward, we have three options:

1. Stick with the option name `preprocessing` instead of `preprocessing_cls` to maintain consistency with the naming of other options.
2. Use the `preprocessing_cls` name for this option and rename other task-specific options as needed.
3. Add a postfix name to all task-specific options. However, this may cause issues with existing integrations, so it might not be the best approach.

I would appreciate your input on which option you prefer. Our goal is to eventually merge this PR, so please let me know what is currently blocking it.

On a side note, I found the experiment you shared here (https://wandb.ai/glenn-jocher/exp-preprocessing) quite interesting, although not unexpected. These options were specifically added to improve performance on datasets with large aspect ratios, which is not the case with ImageNet.

Let me know your preferred option, and we can work towards resolving the blocker and merging the PR.

**Comment by timstokman** [comment]
@glenn-jocher The current PR has already implemented "option 2" as suggested earlier by you. This option is perfectly fine and workable. My preference is towards that option, if you have a different preference, please let me know and I'll implement it.

The PR has also been [tested](https://github.com/ultralytics/ultralytics/pull/3116#issuecomment-1586700671). Both manually, and through the added unit tests.

If possible, I would like to move towards approving/merging this. Is there anything still here blocking an approval/merge?

**Comment by timstokman** [comment]
@glenn-jocher I've simplified this PR a ton. Could you take another look, see if this can be merged?

**Comment by glenn-jocher** [comment]
@timstokman, thank you for your continued effort and initiative to make sure this PR is aligned with the repository's standards and practices. Your willingness to adjust and simplify the PR is greatly appreciated.

Regarding the option naming convention, if you have already aligned with "option 2" where `preprocessing_cls` indicates it's specific to the classification task and you've also adjusted other task-specific options accordingly for consistency, I think that direction is clear and maintains coherence across the codebase.

Furthermore, your thorough testing, both manual and through unit tests, is essential and provides confidence in the changes introduced by your PR.

Since you've simplified the PR and ensured that it adheres to our naming conventions, tests well, and has passed all checks, I don't foresee any additional blockers to approving and merging it. Your cooperation in addressing the previous concerns should pave the way for a smoother integration of your contribution.

Please give me some time to review your latest changes. If everything is in order, we will proceed towards the approval and merging process. Again, thank you for your contributions and patience throughout this process. Your work helps us all progress towards a better and more flexible toolset.

Keep an eye out for any further comments or requests for changes that may arise during the final review. Your prompt and attentive responses have been crucial so far, and we're nearing the completion of this journey.

Let's aim to bring this to a successful conclusion. Thank you for being a valuable member of the Ultralytics community!

**Comment by timstokman** [comment]
> Regarding the option naming convention, if you have already aligned with "option 2" where preprocessing_cls indicates it's specific to the classification task and you've also adjusted other task-specific options accordingly for consistency, I think that direction is clear and maintains coherence across the codebase.

Minor correction, after the introduction of the `rect` option in ultralytics (which happened after this PR was opened), I changed the naming convention in this PR to option 1 because

* The option I introduced had some interactions with the `rect` option
* This was the simplest and most consistent way of resolving that

I also see there are some minor test case failures, those are due to youtube downloads failing, should be fine after a restart.

**Comment by glenn-jocher** [comment]
@timstokman, thank you for the update and clarification.

It seems that you've opted for option 1, naming the option `preprocessing`, which aligns well with the existing convention used by other options like `rect`. This approach keeps the interface simple for users, avoiding potential confusion with task-specific naming.

I understand your decision, especially considering the interaction with the `rect` option, and I appreciate your focus on consistency across the codebase. This decision should indeed help maintain a clean API for users engaging with our command-line tools.

Regarding the test case failures due to YouTube downloads, it's not uncommon for such errors to occur from time to time due to external dependencies. Restarting the tests usually resolves these transient issues. Please proceed to do so in order to have a green signal in all our CI checks.

Provided that all checks pass after restarting the failed ones, and there are no further changes or concerns, I'm inclined to move forward with the merge process. It's always a pleasure to see contributors like you proactively and thoughtfully engage with the development process to foster enhancements that benefit the whole community.

Please feel free to push any new commits if needed to trigger the CI checks again, and I will review the changes closely once all tests provide a clear result. We are indeed close to finalizing this PR, and your dedication toward improving this repository is commendable.

Thank you again for your diligence and contribution! We're excited to see this feature become part of the ultralytics ecosystem. Keep up the great work!

**Comment by timstokman** [comment]
@glenn-jocher All the checks have passed now

**Comment by glenn-jocher** [comment]
@timstokman, excellent news! With all checks having passed, we can proceed towards finalizing your PR.

I will review the changes once more to ensure everything aligns with the repository's guidelines and overall architecture. If no further issues are found, the PR will be approved, and we can go ahead with the merge.

Once again, thank you for your patience, thorough testing, and quality assurance throughout this process. It's contributions like yours that drive innovation and continuous improvement in our collaborative projects.

I'll make sure to update you as soon as the review is complete and provide any additional information regarding the merging status. Your dedication to contributing high-quality code is greatly appreciated.

Looking forward to incorporating your enhancements into the codebase! 🚀

**Comment by timstokman** [comment]
Closing due to no pull request review. I don't have time to keep refactoring this with each conflict I get from the main branch.

**Comment by glenn-jocher** [comment]
@timstokman i'm sorry to hear about your experience. We truly value your contribution and the time you've invested. I'll review the PR promptly to prevent further delays. Thank you for your patience and understanding. 🙏

**Comment by calibratorterminator55** [comment]
@timstokman Hi Tim!

I hope you're doing great! I encountered the same issue you did and was struggling to find a solution for the CenterCrop problem. In the original yolov8 main repository, there's a hard coding of albumentations and CenterCrop augmentations for classification training.

Given that my custom dataset includes images with different ratios and rectangular shapes, the LetterBox augmentation became essential. Unfortunately, I couldn't implement it myself, but then I stumbled upon your solution and your dedication to this issue. I'm truly grateful for the solution you shared. I applied it from your fork, and it significantly boosted the accuracy of my model.

I sincerely hope others recognize the effort you've put into this matter. Fingers crossed that this PR will be merged in the near future.

Thanks once again! 🌟

**Comment by timstokman** [comment]
@calibratorterminator55 Haven't received a single piece of actual code review feedback despite waiting for 7 months, so I'm not counting on it. Can't afford to keep investing time in this unfortunately. I'm glad my branch was useful to people though, I'll leave it up.
